raster_array/tests/test_raster_array.py

```diff
@@ -66,3 +77,84 @@ def test_raster_array_shape_error(raster_4_x_4_multiband):
         RasterArrayShapeError, match="Array must have 3 dimensions, has 2"
     ):
         RasterArray(array[0], metadata)
+
+
+def test_raster_array_dtype_error(raster_4_x_4_multiband):
+    array = raster_4_x_4_multiband.read()
+    metadata = RasterMetadata.from_profile(raster_4_x_4_multiband.profile)
+    metadata = metadata.copy(dtype="int64")
+
+    with pytest.raises(RasterArrayDtypeError, match="does not match metadata dtype"):
+        RasterArray(array, metadata)
+
+
+# test RasterArray.from_raster -------------------------------------------------
+type_and_nodata_coercion_data = [
+    # (data, src_nodata, target_nodata, src_dtype, target_dtype)
+    ([[[0.0, 1.0], [1.0, 0.0]]], 0, -99, np.float32, np.int16),
+    ([[[0.0, 1.0], [1.0, 0.0]]], 0, 0, np.float32, np.int16),
+    ([[[0.0, 1.0], [1.0, 0.0]]], 0, 0, np.int16, np.float32),
+    ([[[0.0, 1.0], [1.0, 0.0]]], 0, np.nan, np.float32, np.float32),
+    ([[[np.nan, 1.0], [1.0, np.nan]]], np.nan, -99, np.float32, np.float32),
+]
+
+
+@pytest.mark.parametrize(
+    "data, src_nodata, target_nodata, src_dtype, target_dtype",
+    type_and_nodata_coercion_data,
+)
+def test_from_raster_simple_target_nodata_or_dtype_coercion(
+    data, src_nodata, target_nodata, src_dtype, target_dtype
+):
+    array = np.array(data, dtype=src_dtype)
+    expected_array = array.astype(target_dtype, copy=True)
+    replacement_mask = (
+        np.isnan(expected_array)
+        if np.isnan(src_nodata)
+        else expected_array == src_nodata
+    )
+    expected_array[replacement_mask] = target_nodata
+
+    with generate_raster(data, src_nodata, src_dtype) as src:
+        raster = RasterArray.from_raster(
+            src, target_nodata=target_nodata, target_dtype=target_dtype
+        )
+
+        assert np.array_equal(raster.array, expected_array, equal_nan=True)
+        assert raster.array.dtype == target_dtype
+        assert raster.metadata.dtype == target_dtype
+        if np.isnan(target_nodata):
+            assert np.isnan(target_nodata)
+        else:
+            assert raster.metadata.nodata == target_nodata
```

Comment:
This looks dumb, but `np.nan != np.nan` so need the different assertion. Unless you (reviewer) can think of something nicer.